torch/onnx/verify.py

```diff
@@ -0,0 +1,262 @@
+import copy
+import io
+import os
+import tempfile
+from contextlib import ExitStack
+from typing import Optional, Union
+
+import numpy as np
+
+import torch
+from torch import Tensor
+from torch.onnx.utils import unpack_quantized_tensor
+
+_ORT_PROVIDERS = ["CPUExecutionProvider"]
+
+
+def _flatten_tuples(elem):
+    flattened = []
+    for t in elem:
+        if isinstance(t, tuple):
+            flattened.extend(_flatten_tuples(t))
+        else:
+            flattened.append(t)
+    return flattened
+
+
+def _to_numpy(elem):
+    if isinstance(elem, Tensor):
+        if elem.requires_grad:
+            return elem.detach().cpu().numpy()
+        else:
+            return elem.cpu().numpy()
+    elif isinstance(elem, (list, tuple)):
+        return [_to_numpy(inp) for inp in elem]
+    elif isinstance(elem, bool):
+        return np.array(elem, dtype=bool)
+    elif isinstance(elem, int):
+        return np.array(elem, dtype=int)
+    elif isinstance(elem, float):
+        return np.array(elem, dtype=float)
+    elif isinstance(elem, dict):
+        flattened = []
+        for k in elem:
+            flattened += [_to_numpy(k)] + [_to_numpy(elem[k])]
+        return flattened
+    return elem
+
+
+def _convert_to_onnx(
+    model,
+    model_f: Union[str, io.BytesIO],
+    input=None,
+    opset_version=None,
+    do_constant_folding=True,
+    keep_initializers_as_inputs=True,
+    dynamic_axes=None,
+    input_names=None,
+    output_names=None,
+    fixed_batch_size=False,
+    training=None,
+    verbose=False,
+    ort_providers=_ORT_PROVIDERS,
+    ort_optim_on=True,
+):
+    input_copy = copy.deepcopy(input)
+
+    torch.onnx._export(
+        model,
+        input_copy,
+        model_f,
+        opset_version=opset_version,
+        do_constant_folding=do_constant_folding,
+        keep_initializers_as_inputs=keep_initializers_as_inputs,
+        dynamic_axes=dynamic_axes,
+        input_names=input_names,
+        output_names=output_names,
+        fixed_batch_size=fixed_batch_size,
+        training=training,
+        verbose=verbose,
+    )
+
+    try:
+        import onnxruntime  # type: ignore[import]
+    except ImportError:
+        raise RuntimeError("ONNXRuntime is required for export verification.")
+
+    # compute onnxruntime output prediction
+    so = onnxruntime.SessionOptions()
+    so.graph_optimization_level = (
+        onnxruntime.GraphOptimizationLevel.ORT_ENABLE_EXTENDED
+        if ort_optim_on
+        else onnxruntime.GraphOptimizationLevel.ORT_DISABLE_ALL
+    )
+    # suppress ort warnings.
+    # 0:Verbose, 1:Info, 2:Warning. 3:Error, 4:Fatal. Default is 2.
+    so.log_severity_level = 3
+    ort_sess = onnxruntime.InferenceSession(
+        model_f if isinstance(model_f, str) else model_f.getvalue(),
+        so,
+        providers=ort_providers,
+    )
+    return ort_sess
+
+
+def _inline_flatten_list(inputs, res_list):
+    for i in inputs:
```

Comment:
naming: ditto

Reply:
ditto

Reply:
`i` should specifically be a counter, taking numbers as values and most likely be an integer. Here i stands for input, which imo should be spelled out. Very nick picky though. I would do

```
for input_ in inputs:
    ...
```

or `val`, or `elem`.